---
title: "4 - Resistome basics"
output: html_notebook
---
# Number of ARGs per AMR class
```{r}
# Show available ranks in the dataset
rank_names(Final.ps)

# Table to show the number of ARG clusters categorised into each AMR class 
table("tax_table"(ARGcluster.aggl)[, "ARG_class"], exclude = NULL)
# 85 ARGs categorised into 8 AMR classes 
```
# Abundance of ARG clusters across COPD, control and blank groups
```{r}
# New ps with only COPD patients
COPD_ps <- subset_samples(ARGcluster.aggl, copdcaco == "1")
# Access the otu_table and tax_table from the COPD_ps phyloseq object
COPD_otu_table <- otu_table(COPD_ps)
COPD_tax_table <- tax_table(COPD_ps)
# Get the row indices of taxa with at least one non-zero value
COPD_non_zero_taxa <- which(rowSums(COPD_otu_table > 0) > 0)
# Extract the taxa names with ARGCluster information
COPD_ARGCluster_names <- as.character(COPD_tax_table[COPD_non_zero_taxa, "ARGCluster"])
# Print the ARGCluster names
print(COPD_ARGCluster_names)

# Controls 
control_ps <- subset_samples(ARGcluster.aggl, copdcaco == "0")
# Access the otu_table and tax_table from the control_ps phyloseq object
control_otu_table <- otu_table(control_ps)
control_tax_table <- tax_table(control_ps)
# Get the row indices of taxa with at least one non-zero value
control_non_zero_taxa <- which(rowSums(control_otu_table > 0) > 0)
# Extract the taxa names with ARGCluster information
control_ARGCluster_names <- as.character(control_tax_table[control_non_zero_taxa, "ARGCluster"])
# Print the ARGCluster names
print(control_ARGCluster_names)

# Blanks 
blank_ps <- subset_samples(ARGcluster.aggl, copdcaco == "blank")
# Access the otu_table and tax_table from the COPD_ps phyloseq object
blanks_otu_table <- otu_table(blank_ps)
blanks_tax_table <- tax_table(blank_ps)
# Get the row indices of taxa with at least one non-zero value
blanks_non_zero_taxa <- which(rowSums(blanks_otu_table > 0) > 0)
# Extract the taxa names with ARGCluster information
blanks_ARGCluster_names <- as.character(blanks_tax_table[blanks_non_zero_taxa, "ARGCluster"])
# Print the ARGCluster names
print(blanks_ARGCluster_names)

# convert matrices to dataframes
df_copd_ARGclusters <- as.data.frame(COPD_ARGCluster_names)
df_control_ARGclusters <- as.data.frame(control_ARGCluster_names)
df_blank_ARGclusters <- as.data.frame(blanks_ARGCluster_names)

# Export data frames to Excel files
write.xlsx(df_copd_ARGclusters, "Data//processed//Basic_resistome_explorations//df_copd_ARGclusters.xlsx", rowNames = FALSE)
write.xlsx(df_control_ARGclusters, "Output_files//Basic_resistome_explorations//df_control_ARGclusters.xlsx", rowNames = FALSE)
write.xlsx(df_blank_ARGclusters, "Output_files//Basic_resistome_explorations//df_blank_ARGclusters.xlsx", rowNames = FALSE)
```
## 16S qPCR DNA load across COPD, control and blank groups
```{r}
# this ps object is clustered at the 90% identity level and blanks removed
ARGcluster.aggl.noblanks 

# Subset the phyloseq object based on 'copdcaco' metadata column
copd_data <- subset_samples(ARGcluster.aggl.noblanks, copdcaco == 1)
control_data <- subset_samples(ARGcluster.aggl.noblanks, copdcaco == 0)

# Calculate summary statistics for 'qPCR_16S_ngml' for COPD patients and controls
copd_qPCR16S_summary <- summary(copd_data@sam_data$qPCR_16S_ngml)
control_qPCR16S_summary <- summary(control_data@sam_data$qPCR_16S_ngml)

# Check distribution in each group
hist(copd_data@sam_data$qPCR_16S_ngml)
hist(control_data@sam_data$qPCR_16S_ngml)
# Distribution in both groups not normal, therefore will use Wilcoxon rank-sum test

# Perform Wilcoxon rank-sum test
wilcox_test_result <- wilcox.test(
  copd_data@sam_data$qPCR_16S_ngml,
  control_data@sam_data$qPCR_16S_ngml
)

# Extract the p-value
p_value <- wilcox_test_result$p.value

# Create boxplot comparing bacterial load between COPD and control groups
plot_data <- data.frame(
  Group = c(rep("COPD", length(copd_data@sam_data$qPCR_16S_ngml)),
            rep("Control", length(control_data@sam_data$qPCR_16S_ngml))),
  Bacterial_Load = c(copd_data@sam_data$qPCR_16S_ngml, control_data@sam_data$qPCR_16S_ngml)
)

# Create the boxplot with p-value annotation
Bacterial_load_comparison <- ggplot(plot_data, aes(x = Group, y = Bacterial_Load, fill = Group)) +
  geom_boxplot() +
  labs(
    title = "Bacterial load (qPCR 16S) comparison between COPD case and control groups",
    x = "Group",
    y = "Bacterial load (qPCR 16S (ngml))",
    caption = paste("Wilcoxon Rank-Sum Test p-value: ", format(p_value, digits = 3))
  ) +
  scale_fill_manual(values = c("cornflowerblue", "coral1")) +  # Reversed colors
  theme_minimal()+
  theme(plot.background = element_rect(fill = "white"))

# Save the ggplot to the output_files folder
ggsave(filename = "Output_files/Bacterial_load_comparison.png", plot = Bacterial_load_comparison, width = 8, height = 6, dpi = 300)
```
## Checking FPK values across COPD, control and blank groups
```{r}
# FPK means not corrected for 16S qPCR values (Fragments Per Kilobase of transcript (FPK))
PS_FPK <- readRDS("Output_files//Phyloseq_objects//3_COPD_resistome_phyloseq_object_rarefied_FPK.RDS") # this is the PS object which is not corrected for 16S qPCR values

# Remove duplicate sample (13674) from the dataset - there was an error in the choice and labelling of 1 control sample and we ended up with the same control sample being used for 2 case samples. It was mislabelled as '13674' hence was thought to be a different sample to '13764' which was already included in the dataset. Therefore we must remoce '13674' from the dataset. 
PS_FPK <- prune_samples(sample_names(PS_FPK) != "13674", PS_FPK) 
PS_FPK

# Now agglomerate the data to ARG cluster level (as above)
# Use tax_glom2 function - #6 is the ARGCluster level in the tax_table of Final.ps object. This creates a new ps object with ARGcluster as default 
ARGcluster.aggl.FPK <- tax_glom2(PS_FPK, taxrank=rank_names(PS_FPK)[6], NArm=TRUE)
ARGcluster.aggl.FPK # Now 85 taxa (as before)

# remove blanks
blanks <- c("veldbl16", "veldbl3", "veldbl5")

ARGcluster.aggl.FPK.noblanks <- ARGcluster.aggl.FPK
ARGcluster.aggl.FPK.noblanks <-  prune_samples(!sample_names(ARGcluster.aggl.FPK.noblanks) 
                                               %in% blanks, ARGcluster.aggl.FPK.noblanks)

# Now can compute FPK per individual 
# Subset the data into COPD and control groups
copd_data_FPK <- subset_samples(ARGcluster.aggl.FPK.noblanks, copdcaco == 1)
control_data_FPK <- subset_samples(ARGcluster.aggl.FPK.noblanks, copdcaco == 0)

# Function to calculate total FPKM for each individual
calculate_total_FPK <- function(physeq_data) {
#' @title Calculates total FPK per sample
#' @description This function takes a phyloseq object 
#' and computes the total FPK per sample in the phyloseq.
#' It then returns the total FPK per sample.
#' @param physeq_data The phyloseq object
#' @return The total FPK value per sample.
  total_FPK <- rowSums(otu_table(physeq_data))
  return(total_FPK)
}
?calculate_total_FPK

# Calculate total FPKM for COPD and control groups
copd_total_FPK <- calculate_total_FPK(copd_data_FPK)
control_total_FPK <- calculate_total_FPK(control_data_FPK)

summary(copd_total_FPK)
summary(control_total_FPK)

hist(copd_total_FPK)
hist(control_total_FPK)
# both not normal distribution - should test with wilcoxon rank sum test

# Create a data frame for boxplot visualization
plot_data_FPK <- data.frame(
  Group = factor(c(rep("COPD", length(copd_total_FPK)), rep("Control", length(control_total_FPK)))),
  Total_FPK = c(copd_total_FPK, control_total_FPK)
)

# Calculate summary statistics for both groups
summary_stats <- summary(plot_data$Total_FPK)

# Create boxplot with summary statistics annotation
FPK_comparison <- ggplot(plot_data_FPK, aes(x = Group, y = Total_FPK, fill = Group)) +
  geom_boxplot() +
  labs(
    title = "Total FPK Comparison between COPD and Control Groups",
    x = "Group",
    y = "Total FPK",
    caption = paste("Wilcoxon Rank-Sum Test p-value:",
                    format(wilcox.test(Total_FPK ~ Group, data = plot_data_FPK)$p.value, digits = 3))
  ) +
  theme_minimal() +
  scale_fill_manual(values = c("COPD" = "coral1", "Control" = "cornflowerblue"))+
  theme(plot.background = element_rect(fill = "white"))

# Save the ggplot to the output_files folder
ggsave(filename = "Output_files/FPK_comparison.png", plot = FPK_comparison, width = 8, height = 6, dpi = 300)
```
# Checking FPKM values
```{r}
# Corrected for 16S qPCR (Fragments Per Kilobase of transcript per Million mapped reads (FPKM))
ARGcluster.aggl.noblanks # this is the fully corrected PS object (FPKM)

# Subset the data into COPD and control groups
copd_data <- subset_samples(ARGcluster.aggl.noblanks, copdcaco == 1)
control_data <- subset_samples(ARGcluster.aggl.noblanks, copdcaco == 0)

# Function to calculate total FPKM for each individual
calculate_total_FPKM <- function(physeq_data) {
  total_FPKM <- rowSums(otu_table(physeq_data))
  return(total_FPKM)
}

# Calculate total FPKM for COPD and control groups
copd_total_FPKM <- calculate_total_FPKM(copd_data)
control_total_FPKM <- calculate_total_FPKM(control_data)
combined_total_FPKM <- calculate_total_FPKM(ARGcluster.aggl.noblanks)

summary(copd_total_FPKM)
summary(control_total_FPKM)
summary(combined_total_FPKM)


hist(copd_total_FPKM)
hist(control_total_FPKM)
hist(combined_total_FPKM)

# both not normal distibution - should test with wilcoxon rank sum test

# Create a data frame for boxplot visualization
plot_data_FPKM <- data.frame(
  Group = factor(c(rep("COPD", length(copd_total_FPKM)), rep("Control", length(control_total_FPKM)))),
  Total_FPKM = c(copd_total_FPKM, control_total_FPKM)
)

# Calculate summary statistics for both groups
summary_stats <- summary(plot_data_FPKM$Total_FPKM)

# Create boxplot with summary statistics annotation
FPKM_comparison <- ggplot(plot_data_FPKM, aes(x = Group, y = Total_FPKM, fill = Group)) +
  geom_boxplot() +
  labs(
    title = "Total FPKM Comparison between COPD and Control Groups",
    x = "Group",
    y = "Total FPKM",
    caption = paste("Wilcoxon Rank-Sum Test p-value:",
                    format(wilcox.test(Total_FPKM ~ Group, data = plot_data_FPKM)$p.value, digits = 3))
  ) +
  theme_minimal() +
  scale_fill_manual(values = c("COPD" = "coral1", "Control" = "cornflowerblue"))+
  theme(plot.background = element_rect(fill = "white"))

# Save the ggplot to the output_files folder
ggsave(filename = "Output_files/FPKM_comparison.png", plot = FPKM_comparison, width = 8, height = 6, dpi = 300)
```
# Additional checks - FPKM per sample
```{r}

count_below_1 <- sum(combined_total_FPKM < 1) # There are none below 1 
hist(combined_total_FPKM, breaks = 50, main = "Histogram of raw FPKM", xlab = "FPKM")
summary(combined_total_FPKM)

# Only round raw FPKM  
rounded_FPKM <- round(combined_total_FPKM)
hist(rounded_FPKM, breaks = 50, main = "Histogram of 10^9 FPKM", xlab = "FPKM")
summary(rounded_FPKM)

# Only transform raw FPKM 
transformed_FPKM <- combined_total_FPKM * 1e9
transformed_FPKM
hist(transformed_FPKM, breaks = 50, main = "Histogram of 10^9 FPKM", xlab = "FPKM")

# Round then transform the raw FPKM
rounded_FPKM <- round(combined_total_FPKM)
rounded_transformed_FPKM <- rounded_FPKM * 1e9
hist(rounded_transformed_FPKM, breaks = 50, main = "Histogram of rounded then 10^9 transformed FPKM", xlab = "FPKM")
summary(rounded_transformed_FPKM)

# Transform then round the raw FPKM
transformed_FPKM <- combined_total_FPKM * 1e9
transformed_rounded_FPKM <- round(transformed_FPKM)
hist(transformed_rounded_FPKM, breaks = 50, main = "Histogram of 10^9 transformed, then rounded FPKM", xlab = "FPKM")
summary(transformed_rounded_FPKM)
```
# Additional checks - FPKM per ARG
```{r}
# Extract the OTU table
ARG_table <- as.data.frame(otu_table(ARGcluster.aggl.noblanks))

# Total FPKM per ARG
## Raw FPKM
total_fpkm_per_arg <- colSums(ARG_table)
arg_fpkm_data <- data.frame(Total_FPKM = total_fpkm_per_arg)
print(arg_fpkm_data)
hist(arg_fpkm_data$Total_FPKM, breaks = 30, col = "skyblue", main = "Distribution of ARG Total FPKMs", xlab = "Total FPKM")
min(arg_fpkm_data$Total_FPKM)
median(arg_fpkm_data$Total_FPKM)
mean(arg_fpkm_data$Total_FPKM)
max(arg_fpkm_data$Total_FPKM)

arg_below_1 <- arg_fpkm_data[arg_fpkm_data$Total_FPKM < 1, ]

# zoom in on low FPKM values (up to total FPKM of 100)
subset_data_0to100 <- arg_fpkm_data[arg_fpkm_data$Total_FPKM >= 0 & arg_fpkm_data$Total_FPKM <= 100, , drop = FALSE]
subset_data_0to100$ARG <- rownames(subset_data_0to100)
rownames(subset_data_0to100) <- NULL
hist(subset_data_0to100$Total_FPKM, breaks = 30, col = "skyblue", main = "Distribution of ARG Total FPKMs (0-100)", xlab = "Total FPKM")
plot(subset_data_0to100$ARG, subset_data_0to100$Total_FPKM, pch = 16, col = "skyblue", main = "Individual FPKM Distribution for ARGs (0-100)", xlab = "ARG", ylab = "Total FPKM")
ggplot(subset_data_0to100, aes(x = ARG, y = Total_FPKM)) +
  geom_point(aes(color = Total_FPKM), size = 3) +
  scale_color_gradient(low = "skyblue", high = "darkblue") +
  labs(title = "Total FPKM values per ARG (for ARGs with a total FPKM 0-100)", x = "", y = "Total FPKM") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))





## Round then transform the raw FPKM per ARG
ARG_table_rounded <- as.data.frame(round(otu_table(ARGcluster.aggl.noblanks)))
ARG_table_rounded_transformed <-  ARG_table_rounded* 1e9
total_fpkm_per_arg_rounded_transformed <- colSums(ARG_table_rounded_transformed)
arg_fpkm_data_rounded_transformed <- data.frame(Total_FPKM_rounded_transformed = total_fpkm_per_arg_rounded_transformed)
print(arg_fpkm_data_rounded_transformed)
hist(arg_fpkm_data_rounded_transformed$Total_FPKM, breaks = 30, col = "skyblue", main = "Distribution of ARG Total FPKMs (rounded then 10^9 transformed)", xlab = "Total FPKM")
summary(arg_fpkm_data_rounded_transformed)

min(arg_fpkm_data_rounded_transformed$Total_FPKM)
median(arg_fpkm_data_rounded_transformed$Total_FPKM)
mean(arg_fpkm_data_rounded_transformed$Total_FPKM)
max(arg_fpkm_data_rounded_transformed$Total_FPKM)

## Transform then round the raw FPKM
ARG_table_transformed <- as.data.frame(otu_table(ARGcluster.aggl.noblanks)* 1e9)
ARG_table_transformed_rounded <- round(ARG_table_transformed)
total_fpkm_per_arg_transformed_rounded <- colSums(ARG_table_transformed_rounded)
arg_fpkm_data_transformed_rounded <- data.frame(Total_FPKM_transformed_rounded = total_fpkm_per_arg_transformed_rounded)
print(arg_fpkm_data_transformed_rounded)
hist(arg_fpkm_data_transformed_rounded$Total_FPKM_transformed_rounded, breaks = 30, col = "skyblue", main = "Distribution of ARG Total FPKMs (10^9 transformed then rounded)", xlab = "Total FPKM")
summary(arg_fpkm_data_transformed_rounded)

min(arg_fpkm_data_transformed_rounded$Total_FPKM_transformed_rounded)
median(arg_fpkm_data_transformed_rounded$Total_FPKM_transformed_rounded)
mean(arg_fpkm_data_transformed_rounded$Total_FPKM_transformed_rounded)
max(arg_fpkm_data_transformed_rounded$Total_FPKM_transformed_rounded)





```

# Rarefaction curves
```{r}
# Since there are no FPKM values <1, instead of the multiply by billion method, we can just round for alpha diversity calculations
Final.ps.round <- Final.ps
Final.ps.round@otu_table@.Data <- round(Final.ps.round@otu_table@.Data) 

# rename the copdcaco categories
Final.ps.round@sam_data$copdcaco <- revalue(Final.ps.round@sam_data$copdcaco, c('1' ='COPD', '0' = 'control', 'blank'= 'blank'))

# Rarefaction curves on rounded data
# Create rarefraction function (adapted from vegan `rarecurve` function)
ggrare <- function(physeq_object, step = 10, label = NULL, color = NULL, plot = TRUE, parallel = FALSE, se = TRUE) {

  x <- methods::as(phyloseq::otu_table(physeq_object), "matrix")
  if (phyloseq::taxa_are_rows(physeq_object)) { x <- t(x) }

  tot <- rowSums(x)
  S <- rowSums(x > 0)
  nr <- nrow(x)

  rarefun <- function(i) {
    cat(paste("rarefying sample", rownames(x)[i]), sep = "\n")
    n <- seq(1, tot[i], by = step)
    if (n[length(n)] != tot[i]) {
      n <- c(n, tot[i])
    }
    y <- vegan::rarefy(x[i, ,drop = FALSE], n, se = se)
    if (nrow(y) != 1) {
      rownames(y) <- c(".S", ".se")
      return(data.frame(t(y), Size = n, Sample = rownames(x)[i]))
    } else {
      return(data.frame(.S = y[1, ], Size = n, Sample = rownames(x)[i]))
    }
  }
  if (parallel) {
    out <- parallel::mclapply(seq_len(nr), rarefun, mc.preschedule = FALSE)
  } else {
    out <- lapply(seq_len(nr), rarefun)
  }
  df <- do.call(rbind, out)

  # Get sample data
  if (!is.null(phyloseq::sample_data(physeq_object, FALSE))) {
    sdf <- methods::as(phyloseq::sample_data(physeq_object), "data.frame")
    sdf$Sample <- rownames(sdf)
    data <- merge(df, sdf, by = "Sample")
    labels <- data.frame(x = tot, y = S, Sample = rownames(x))
    labels <- merge(labels, sdf, by = "Sample")
  }

  # Add, any custom-supplied plot-mapped variables
  if ( length(color) > 1 ) {
    data$color <- color
    names(data)[names(data) == "color"] <- deparse(substitute(color))
    color <- deparse(substitute(color))
  }

  if ( length(label) > 1 ) {
    labels$label <- label
    names(labels)[names(labels) == "label"] <- deparse(substitute(label))
    label <- deparse(substitute(label))
  }

  p <- ggplot2::ggplot(data = data,
                       ggplot2::aes_string(x = "Size",
                                           y = ".S",
                                           group = "Sample",
                                           color = color))

  p <- p + ggplot2::labs(x = "Sequence Sample Size", y = "Species Richness")

  if (!is.null(label)) {
    p <- p + ggplot2::geom_text(data = labels,
                                ggplot2::aes_string(x = "x",
                                                    y = "y",
                                                    label = label,
                                                    color = color),
                                size = 4, hjust = 0)
  }

  p <- p + ggplot2::geom_line()
  if (se) { ## add standard error if available
    p <- p +
      ggplot2::geom_ribbon(ggplot2::aes_string(ymin = ".S - .se",
                                               ymax = ".S + .se",
                                               color = NULL,
                                               fill = color),
                           alpha = 0.2)
  }
  if (plot) {
    plot(p)
  }
  invisible(p)
}

# Compute minimum sequencing depth
min_depth <- min(sample_sums(Final.ps.round)) # This is the minimum sample count achieved over all the samples. We will rarefy the sample counts to this value = 202

# Now make rarefaction curves
resistome_rare_curves <- ggrare(Final.ps.round, step = 1000, color = "copdcaco", se = FALSE) # step = 1000 specifies the interval at which rarefaction is calculated.
# Add cut off line at minimum sequencing depth 
rare_curves_color_copdcaco_split <- resistome_rare_curves + facet_wrap(~copdcaco)

# Save the rarefaction curves
ggsave("Output_files//Rarefaction_curves//Rarefaction_curves_colourperpopulation.png", plot = rare_curves_colorperpopulation, dpi = 300)
rounded_FPKM
```
## Top 10 most abundant ARG clusters
```{r}
# Function to calculate top 10 taxa at different levels of taxonomic identification e.g. at ARG level, ARG cluster level, AMR class level etc. 
toptaxa <- function( ps, rank="ARGCluster", top=10, samples=NA, NArm=FALSE ) {
  if( ! is.na(samples[1]) ) {
    pstemp = prune_samples( samples, ps)
  } else {
    pstemp = ps
  }
  pstemp = tax_glom2( pstemp, taxrank = rank, NArm=NArm )
  pstemp = transform_sample_counts( pstemp, function(x) 100*x/sum(x) )

  if( ntaxa(pstemp) < top ) {
    top = ntaxa(pstemp)
    warning("Number of taxa in object is less then requested top list! Reduced to ", top)
  }

  toplist = names( sort( taxa_sums(pstemp), decreasing = TRUE ) )[1:top]
  return( toplist )
}

top10.ARGclusters.overall<- toptaxa (ARGcluster.aggl, rank="ARGCluster", top=10, samples=NA, NArm=FALSE )
top10.ARGclusters.overall

top10.ARGclusters.overall.noblanks<- toptaxa (ARGcluster.aggl.noblanks, rank="ARGCluster", top=10, samples=NA, NArm=FALSE )
top10.ARGclusters.overall.noblanks

# Top 10 in copd samples 
top10.ARGclusters.COPD<- toptaxa (COPD_ps, rank="ARGCluster", top=10, samples=NA, NArm=FALSE )
top10.ARGclusters.COPD

# Top 10 in control samples 
top10.ARGclusters.control<- toptaxa (control_ps, rank="ARGCluster", top=10, samples=NA, NArm=FALSE )
top10.ARGclusters.control


# Present as actual ARG cluster names 
tt <- data.frame( tax_table( ARGcluster.aggl ) )
# extract the names of the ARGCluster column where the row names of tt match the taxa present in the top10.ARGclusters.noblanks, top10.ARGclusters.COPD and top10.ARGclusters.control

tt$ARGCluster[ row.names(tt) %in% top10.ARGclusters.noblanks] 
tt$ARGCluster[ row.names(tt) %in% top10.ARGclusters.COPD]
tt$ARGCluster[ row.names(tt) %in% top10.ARGclusters.control]


# Create a new ps object which holds only the 10 most abundant ARGs (for all COPD and control samples)
Final.ps.top10ARGClusters <- prune_taxa(names(top10.ARGclusters.noblanks), AMRcluster90.aggl)

#BARPLOT
# Make a barplot like the AMR class level one done above, so that for COPD case and control groups, I present the relative abundance of each of the top 10 ARG classes
bar.graphs2 <- function(physeq, var1 = "ungrouped", level, displayx = 10, filetype = ".jpg") {
  if (level == "Species") {
    print("Error this level is too fine for amplicon sequencing")
  } else if (var1 == "ungrouped") {
    physeqxu <- tax_glom(physeq, level)
    physeqxu <- transform_sample_counts(physeqxu, function(x) 100 * x/sum(x))
    physeqxu <- prune_taxa(names(sort(taxa_sums(physeqxu), TRUE))[1:displayx], physeqxu)
    barchartxu <- plot_bar(physeqxu, fill = level) +
      labs(x = "COPD or control samples",
           y = "% of top 10 ARG clusters present",
           title = paste("Top", displayx, level, "ungrouped")) +
      scale_fill_brewer(palette = "Spectral")
    return(barchartxu)
  } else {
    physeqm <- merge_samples(physeq, var1)
    sample_data(physeqm)[[var1]] <- levels(sample_data(physeqm)[[var1]])
    physeqg <- tax_glom(physeqm, level)
    physeqg <- transform_sample_counts(physeqg, function(x) 100 * x/sum(x))
    physeqg <- prune_taxa(names(sort(taxa_sums(physeqg), TRUE))[1:displayx], physeqg)
    grouped_bar <- plot_bar(physeqg, fill = level) +
      labs(x = "COPD or control samples",
           y = "% of top 10 ARG clusters present",
           title = paste("Top", displayx, level, sep = " ")) +
      scale_fill_brewer(palette = "Spectral")
    return(grouped_bar)
  }
}

# Subset the phyloseq object to include only COPD and control samples
ARGcluster.aggl_subset <- subset_samples(AMRcluster90.aggl, copdcaco %in% c("1", "0"))

# Create a new phyloseq object with only the top 10 ARG clusters
ARGcluster.top10 <- prune_taxa(names(sort(taxa_sums(ARGcluster.aggl_subset), decreasing = TRUE))[1:10], ARGcluster.aggl_subset)

# Generate the stacked barplot at the top 10 ARG cluster level
ARGcluster.barchart <- bar.graphs2(ARGcluster.top10, "copdcaco", "ARGCluster") +
  ggtitle("Relative Abundance of Top 10 ARG Clusters in COPD and Control Groups") +
  labs(fill = "ARG Cluster") +
  theme(axis.text = element_text(size = 15),
        axis.title = element_text(size = 15),
        axis.text.x = element_text(angle = 0, hjust = 0.5),  # Align labels to the center of tick points
        title = element_text(size = 17),
        legend.title = element_text(size = 15),
        legend.text = element_text(size = 15),
        legend.position = "right") +
  scale_x_discrete(labels = c("COPD" = "COPD", "control" = "Control"))

ARGcluster.barchart

ggsave("ARGcluster.barchart.png",width = 7, height = 5,dpi = 300)
```